core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java

```diff
@@ -407,11 +407,10 @@ private ThreadContextStruct putHeaders(Map<String, String> headers) {
             if (headers.isEmpty()) {
                 return this;
             } else {
-                final Map<String, String> newHeaders = new HashMap<>();
+                final Map<String, String> newHeaders = new HashMap<>(this.requestHeaders);
```

Comment:
I just realized I'm unsure about this fix. The `putSingleHeader` below will fail if the key already exists, so anything in requestHeaders can now not be set. I think what we want instead is a loop over requestHeaders after the headers loop, but which uses putIfAbsent?

Reply:
@rjernst I think the fix is good. it's really just a shortcut for a single header. I wonder if we should de-optimize and just iterate over the map and use putHeader(string,string) instead? also the test looks great